lib/BuildSystem/LaneBasedExecutionQueue.cpp

```diff
@@ -281,26 +295,88 @@ class LaneBasedExecutionQueue : public BuildExecutionQueue {
     if (shouldCaptureOutput) {
       // Close the write end of the output pipe.
       ::close(outputPipe[1]);
+      ::close(stderrPipe[1]);
 
-      // Read all the data from the output pipe.
-      while (true) {
-        char buf[4096];
-        ssize_t numBytes = read(outputPipe[0], buf, sizeof(buf));
-        if (numBytes < 0) {
+      // Set the output and stderr pipes to non-blocking.
+      int flags = fcntl(outputPipe[0], F_GETFL, 0);
+      if (fcntl(outputPipe[0], F_SETFL, flags | O_NONBLOCK) < 0) {
+        getDelegate().commandProcessHadError(
+          context.job.getForCommand(), handle,
+          Twine("unable to set the non-blocking flag in stdout pipe (") + strerror(errno) + ")");
+        getDelegate().commandProcessFinished(context.job.getForCommand(), handle,
+                                           -1);
+        return false;
+      }
+      flags = fcntl(stderrPipe[0], F_GETFL, 0);
+      if (fcntl(stderrPipe[0], F_SETFL, flags | O_NONBLOCK)) {
+        getDelegate().commandProcessHadError(
+          context.job.getForCommand(), handle,
+          Twine("unable to set the non-blocking flag in stderr pipe (") + strerror(errno) + ")");
+        getDelegate().commandProcessFinished(context.job.getForCommand(), handle,
+                                           -1);
+        return false;
+      }
+     
+      // We use select() to determine which fd has output ready to be read. Once select()
+      // returns we read whichever (or both) fd has data available in non-blocking mode.
+      // We stop watching a fd once it has read all of the data i.e. read() returns 0.
+      // When all of the fds are done being watched we can exit the loop.
+      bool watchingStdout = true, watchingStderr = true;
+      while(true) {
+        fd_set readSet;
+        FD_ZERO(&readSet);
```

Comment:
I still think you can drop `watchingStdout` and just iterate while the fd_set is non-empty.